tests/convergence/ex1c.cpp

```diff
@@ -0,0 +1,414 @@
+   dim = mesh->Dimension();
+
+   mesh->UniformRefinement();
+
+   // Define a finite element space on the mesh.
+   FiniteElementCollection *fec;
+   if (order > 0)
+   {
+      fec = new H1_FECollection(order, dim);
+   }
+   else if (mesh->GetNodes())
+   {
+      fec = mesh->GetNodes()->OwnFEC();
+      cout << "Using isoparametric FEs: " << fec->Name() << endl;
+   }
+   else
+   {
+      fec = new H1_FECollection(order = 1, dim);
+   }
+   FiniteElementSpace *fespace = new FiniteElementSpace(mesh, fec);
+   cout << "Number of finite element unknowns: "
+        << fespace->GetTrueVSize() << endl;
+
+   prob_kind = (prob)iprob;
+   mfem::ComplexOperator::Convention conv = mfem::ComplexOperator::HERMITIAN;
+
+
+   // Set up the linear form b(.)
+   ComplexLinearForm *b = new ComplexLinearForm(fespace,conv);
+   FunctionCoefficient f_re(f_exact_re);
+   FunctionCoefficient f_im(f_exact_im);
+   b->AddDomainIntegrator(new DomainLFIntegrator(f_re),
+                          new DomainLFIntegrator(f_im));
+   b->real().Vector::operator=(0.0);
+   b->imag().Vector::operator=(0.0);
+
+   // Define the solution vector x as a finite element grid function
+   // corresponding to fespace.
+   ComplexGridFunction x(fespace);
+   ComplexGridFunction x_ex(fespace);
+   FunctionCoefficient u_ex_re(u_exact_re);
+   FunctionCoefficient u_ex_im(u_exact_im);
+   VectorFunctionCoefficient u_grad_coeff_re(dim, u_grad_exact_re);
+   VectorFunctionCoefficient u_grad_coeff_im(dim, u_grad_exact_im);
+   // Set up the sesquilinear form a(.,.)
+   ConstantCoefficient one(1.0);
+   SesquilinearForm *a = new SesquilinearForm(fespace,conv);
+   switch (prob_kind)
+   {
+      case comp :
+         a->AddDomainIntegrator(new DiffusionIntegrator(one),
+                                new DiffusionIntegrator(one));
+         break;
+      case rl :
+         a->AddDomainIntegrator(new DiffusionIntegrator(one), NULL);
+         break;
+      case im :
+         a->AddDomainIntegrator(NULL,new DiffusionIntegrator(one));
+         break;
+   }
+
+   double L2err_re0, L2err_im0;
+   double H1err_re0, H1err_im0;
+   for (int l=0; l<ref_levels+1; l++)
+   {
+      a->Assemble();
+      b->Assemble();
+      OperatorHandle Ah;
+      Vector B, X;
+      // Determine the list of true (i.e. conforming) essential boundary dofs.
+      Array<int> ess_bdr, ess_tdof_list;
+      if (mesh->bdr_attributes.Size())
+      {
+         ess_bdr.SetSize(mesh->bdr_attributes.Max());
+         ess_bdr = 1;
+         fespace->GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+      }
+      x.ProjectBdrCoefficient(u_ex_re,u_ex_im,ess_bdr);
+      x_ex.ProjectCoefficient(u_ex_re,u_ex_im);
+
+      a->FormLinearSystem(ess_tdof_list, x, *b, Ah, X, B);
+
+      ComplexSparseMatrix * AZ = Ah.As<ComplexSparseMatrix>();
+      SparseMatrix * A = AZ->GetSystemMatrix();
+
+      cout << "Size of linear system: " << A->Height() << endl;
+
+
+#ifndef MFEM_USE_SUITESPARSE
+      {
+         GMRESSolver gmres;
```

Comment:
Adding the `GSSmoother` as a preconditioner, as in `ex1.cpp`, seems to improve performance.  We should try to make sure that this solve converges for each recommended test case for orders up to 3 perhaps.

Reply:
I added Block Diagonal GSSmoother and the tests pass now except the case of a purely imaginary problem. In the parallel case I added a block diagonal BoomerAMG preconditioner and all the tests pass (at least upto order 3 and with two refinements)